[REF] phantom test to webiste_sale_delivery feature in order to test the selection of carriers

hugho-ad opened the pull request:
Running the new phantom tests  in this pr from the web browser, triggers the next error:

to run phantom test, enter the next in to console of web browser:
`openerp.Tour.run('shop_buy_delivery')`
https://youtu.be/I-DpJdMvS1k

In log you will get the next error:

``` python
2016-01-14 17:43:53,330 2564 INFO odoo werkzeug: 172.17.0.1 - - [14/Jan/2016 17:43:53] "GET /shop/payment HTTP/1.1" 200 -
2016-01-14 17:43:53,330 2564 ERROR odoo openerp.sql_db: bad query: UPDATE "sale_order" SET "payment_acquirer_id"=1,"payment_tx_id"=15,"write_uid"=1,"write_date"=(now() at time zone 'UTC') WHERE id IN (17)
Traceback (most recent call last):
  File "/.repo_requirements/odoo/openerp/sql_db.py", line 234, in execute
    res = self._obj.execute(query, params)
TransactionRollbackError: could not serialize access due to concurrent update

```

Then of this continue with the next:

``` python
2016-01-14 17:43:53,331 2564 INFO odoo openerp.service.model: SERIALIZATION_FAILURE, retry 1/5 in 0.6895 sec...
2016-01-14 17:43:53,925 2564 INFO odoo werkzeug: 172.17.0.1 - - [14/Jan/2016 17:43:53] "POST /shop/payment/transaction/1 HTTP/1.1" 200 -
2016-01-14 17:43:53,934 2564 INFO odoo openerp.addons.payment_transfer.controllers.main: Beginning form_feedback with post data {'amount': u'320.0',
 'currency': u'EUR',
 'reference': u'SO016',
 'return_url': u'/shop/payment/validate'}
2016-01-14 17:43:53,939 2564 INFO odoo openerp.addons.payment_transfer.models.payment_acquirer: Validated transfer payment for tx SO016: set as pending
2016-01-14 17:43:54,125 2564 ERROR odoo openerp.http: Exception during JSON request handling.
Traceback (most recent call last):
  File "/.repo_requirements/odoo/openerp/http.py", line 537, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/.repo_requirements/odoo/openerp/http.py", line 574, in dispatch
    result = self._call_function(**self.params)
  File "/.repo_requirements/odoo/openerp/http.py", line 310, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/.repo_requirements/odoo/openerp/service/model.py", line 118, in wrapper
    return f(dbname, *args, **kwargs)
  File "/.repo_requirements/odoo/openerp/http.py", line 307, in checked_call
    return self.endpoint(*a, **kw)
  File "/.repo_requirements/odoo/openerp/http.py", line 803, in __call__
    return self.method(*args, **kw)
  File "/.repo_requirements/odoo/openerp/http.py", line 403, in response_wrap
    response = f(*args, **kw)
  File "/.repo_requirements/odoo/addons/website_sale/controllers/main.py", line 753, in payment_transaction
    }, context=context)
  File "/.repo_requirements/odoo/openerp/api.py", line 256, in wrapper
    return old_api(self, *args, **kwargs)
  File "/.repo_requirements/odoo/addons/payment/models/payment_acquirer.py", line 414, in create
    return super(PaymentTransaction, self).create(cr, uid, values, context=context)
  File "/.repo_requirements/odoo/openerp/api.py", line 256, in wrapper
    return old_api(self, *args, **kwargs)
  File "/.repo_requirements/odoo/addons/mail/mail_thread.py", line 381, in create
    thread_id = super(mail_thread, self).create(cr, uid, values, context=context)
  File "/.repo_requirements/odoo/openerp/api.py", line 256, in wrapper
    return old_api(self, *args, **kwargs)
  File "/.repo_requirements/odoo/openerp/api.py", line 360, in old_api
    result = method(recs, *args, **kwargs)
  File "/.repo_requirements/odoo/openerp/models.py", line 4094, in create
    record = self.browse(self._create(old_vals))
  File "/.repo_requirements/odoo/openerp/api.py", line 254, in wrapper
    return new_api(self, *args, **kwargs)
  File "/.repo_requirements/odoo/openerp/api.py", line 496, in new_api
    result = method(self._model, cr, uid, *args, **old_kwargs)
  File "/.repo_requirements/odoo/openerp/models.py", line 4285, in _create
    recs._validate_fields(vals)
  File "/.repo_requirements/odoo/openerp/api.py", line 254, in wrapper
    return new_api(self, *args, **kwargs)
  File "/.repo_requirements/odoo/openerp/models.py", line 1272, in _validate_fields
    raise ValidationError('\n'.join(errors))
ValidationError: ('ValidateError', u'Field(s) `reference, state` failed against a constraint: The payment transaction reference must be unique!')
2016-01-14 17:43:54,132 2564 INFO odoo werkzeug: 172.17.0.1 - - [14/Jan/2016 17:43:54] "POST /shop/payment/transaction/1 HTTP/1.1" 200 -
2016-01-14 17:43:54,295 2564 INFO ? werkzeug: 127.0.0.1 - - [14/Jan/2016 17:43:54] "GET /web/static/lib/fontawesome/css/font-awesome.css HTTP/1.1
```

**Steps to reproduce**
From website,
/shop
- 'step 0 select ipod',
- 'step 1 select ipod 32GB'
- 'step 2 click on add to cart on ipod details'
- 'step 3 search an ipad in order to have two products',
- 'step 4 go in the ipad information'
- "step 5 add to chart from the full view for ipad"
- "click in pop-up on 'Proceed to checkout' button"
- "step 6 go to checkout"
- "step 7fill the checkout form"
- "step 8 select a delivery method"
- "step 9 select payment"
- "step 10 click onPay Now"

beledouxdenis commented:
I am not sure to get the point of your proposal. 

 - Have you made these tests to show an issue within the standard modules ? If so, can you explain the actual issue you want solved ?
or
 - Do you want to add a series of tests, and you do not know why these are failing ?

Regards,

hugho-ad commented:
Hi @beledouxdenis 

My proposal is add a phantom test, and show you an issue  within the standard modules when running the phantom test that I added in this pr. I don't know why these are failing. Can you help me with this?

while debugging we find that each request creates a diferrent cursor:
Debug:
 - Added follow prints sentence:
`import traceback;traceback.print_stack();print "force_carrier_id", force_carrier_id, "cr", cr, "order.id", order.id` in follow file `/root/odoo-8.0/addons/website_sale_delivery/models/sale_order.py` in line 85, in method `_check_carrier_quotation`

 - I ran steps again and saw follow result:
```bash
File "/root/odoo-8.0/addons/website_sale_options/controllers/main.py", line 46, in cart_options_update_json
    optional_product_ids=optional_product_ids)
  File "/root/odoo-8.0/addons/website_sale_delivery/models/sale_order.py", line 141, in _cart_update
    self._check_carrier_quotation(cr, uid, sale_order, context=context)
  File "/root/odoo-8.0/addons/website_sale_delivery/models/sale_order.py", line 86, in _check_carrier_quotation
force_carrier_id None cr <openerp.sql_db.Cursor object at 0x7f08b01d32d0> order.id 151
...
File "/root/odoo-8.0/addons/website_sale/controllers/main.py", line 648, in confirm_order
    request.website.sale_get_order(update_pricelist=True, context=context)
  File "/root/odoo-8.0/addons/website_sale/models/sale_order.py", line 197, in sale_get_order
    sale_order._cart_update(product_id=line.product_id.id, line_id=line.id, add_qty=0)
  File "/root/odoo-8.0/addons/website_sale_delivery/models/sale_order.py", line 141, in _cart_update
    self._check_carrier_quotation(cr, uid, sale_order, context=context)
  File "/root/odoo-8.0/addons/website_sale_delivery/models/sale_order.py", line 86, in _check_carrier_quotation
force_carrier_id None cr <openerp.sql_db.Cursor object at 0x7f08a3dbc310> order.id 151
...
File "/root/odoo-8.0/addons/website_sale/controllers/main.py", line 648, in confirm_order
    request.website.sale_get_order(update_pricelist=True, context=context)
  File "/root/odoo-8.0/addons/website_sale/models/sale_order.py", line 197, in sale_get_order
    sale_order._cart_update(product_id=line.product_id.id, line_id=line.id, add_qty=0)
  File "/root/odoo-8.0/addons/website_sale_delivery/models/sale_order.py", line 141, in _cart_update
    self._check_carrier_quotation(cr, uid, sale_order, context=context)
  File "/root/odoo-8.0/addons/website_sale_delivery/models/sale_order.py", line 86, in _check_carrier_quotation
force_carrier_id None cr <openerp.sql_db.Cursor object at 0x7f08b0bf1c10> order.id 151
...
 File "/root/odoo-8.0/addons/website_sale_delivery/controllers/main.py", line 18, in payment
    request.registry['sale.order']._check_carrier_quotation(cr, uid, order, force_carrier_id=carrier_id, context=context)
force_carrier_id 4 cr <openerp.sql_db.Cursor object at 0x7f08b0fa7090> order.id 151
2016-01-13 03:47:56,520 14248 ERROR openerp_test openerp.sql_db: bad query: UPDATE "sale_order" SET "carrier_id"=4,"write_uid"=1,"write_dat
e"=(now() at time zone 'UTC') WHERE id IN (151)
Traceback (most recent call last):
  File "/.repo_requirements/odoo/openerp/sql_db.py", line 234, in execute
    res = self._obj.execute(query, params)
TransactionRollbackError: could not serialize access due to concurrent update
...
```
Note:
This [step](https://github.com/vauxoo-dev/odoo/blob/8.0-issue-cr/addons/website_sale_delivery/static/src/js/website_sale.test.js#L39-L42) in test phantom, should not to be here, because  when installing just webiste_sale_delivery, the e-comerce front end doesn't look like instaces in runbot, so that step is a temporal step in order to that the test be able to running

hugho-ad commented:
@beledouxdenis any news?

do you need something else ?

Regards!
